### PR TITLE
[FEATURE] Ensure `result_format` accessed is through Checkpoint, and warns users if `Expectation` or `Validator`-level

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
# What has changed? 
* We prefer `Checkpoints` to configure `result_format` including `unexpected_index_column_names`.
* Configuring `unexpected_index_column_names` At the `Validator` or `ExpectationConfiguration` level will give a warning. 
* The check is facilitated by `_warn_if_result_format_config_in_expectation_configuration()`, and `_warn_if_result_format_config_in_runtime_configuration()` which are helper methods that have been added to Expectation's `validate()` method.

### Tests have been updated and refactored to reflect this change. 
* Tests that were previously in `test_map_metric.py` have been migrated to `test_checkpoint_result_format.py` and now use a `Checkpoint` configuration. 
* Test has been added to `test_map_metric.py` to check that appropriate warning is given when configuring `result_format` at the `ExpectationConfiguration`-level
* Tests have been added to `test_validator.py` to check that appropriate warning is given when configuring `result_format` at the `Validator`-level
* Metrics that are needed for ID/PK, namely `unexpected_index_list` and `unexpected_index_query` are now tested as unittests in `tests/expectations/metrics/column_map_metrics/test_unexpected_indices_and_query.py` 

### Appendix
* Some example configurations to demonstrate what will and not work as the result of this PR. 
#### The following configurations at the `Checkpoint`-level will work
1. in `CheckpointConfiguration`
   
```python
# if Checkpoint configuration contains the following `runtime_configuration`
checkpoint_dict["runtime_configuration"] = {
        "result_format": {
            "result_format": "COMPLETE",
            "unexpected_index_column_names": ["pk_1"],
        }
    }
```

2. In `context.run_checkpoint()`

```python 
result_format: dict = {
    "result_format": "COMPLETE",
    "unexpected_index_column_names": ["pk_1"],
}
result: CheckpointResult = context.run_checkpoint(
    checkpoint_name="my_checkpoint", result_format=result_format
)
```

#### The following configurations at the `Validator`-level and `Expectation`-level will issue warnings.
1. at `validator.expect()`
```python 
# will raise UserWarning
result_format_config: dict = {
        "result_format": "COMPLETE",
        "unexpected_index_column_names": ["pk_1"],
}

result: ExpectationValidationResult = validator.expect_column_values_to_be_in_set(
    column="animals",
    value_set=["cat", "fish", "dog"],
    result_format=result_format_config,
)
```

2. at `expectation.validate()`

```python
# will raise UserWarning
runtime_configuration: dict = {
    "result_format": {
        "result_format": "SUMMARY",
        "unexpected_index_column_names": ["pk_1"],
    }
}
result: ExpectationValidationResult = expectation.validate(
    validator=validator, runtime_configuration=runtime_configuration
)
```

3. At `ExpectationConfiguration`
```python 
# will raise UserWarning
    expectation_configuration = ExpectationConfiguration(
        expectation_type="expect_column_values_to_be_in_set",
        kwargs={
            "column": "animals",
            "value_set": ["cat", "fish", "dog"],
            "result_format": {
                "result_format": "COMPLETE",
                "unexpected_index_column_names": ["pk_1"],
            },
        },
    )
```

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/checkpoint/test_checkpoint_result_format.py [hunk 1]
```diff
@@ -18,62 +18,6 @@
 logger = logging.getLogger(__name__)
 
 
-@pytest.fixture()
-def reference_checkpoint_config_for_unexpected_column_names() -> dict:
```
Comment (Shinnnyshinshin) [inline tests/checkpoint/test_checkpoint_result_format.py:22]: Moved to `conftest.py`

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 2]
```diff
@@ -2441,7 +2441,7 @@ def _sqlalchemy_map_condition_query(
 
 def _sqlalchemy_map_condition_index(
     cls,
-    execution_engine: PandasExecutionEngine,
+    execution_engine: SqlAlchemyExecutionEngine,
```
Comment (Shinnnyshinshin) [inline great_expectations/expectations/metrics/map_metric_provider.py:2444]: oops

File: tests/expectations/core/test_expect_column_values_to_be_decreasing.py [hunk 3]
```diff
@@ -41,5 +41,5 @@ def test_expect_column_values_to_be_decreasing_warn_parse_strings_as_datetimes(
             column="a", parse_strings_as_datetimes=True
         )
     assert 'The parameter "parse_strings_as_datetimes" is deprecated' in str(
-        record.list[0].message
+        record.list[1].message
```
Comment (Shinnnyshinshin) [inline tests/expectations/core/test_expect_column_values_to_be_decreasing.py:44]: Adjusting of warning is needed now that we output one for Expectation and Validator-level `result_format` configurations

File: great_expectations/expectations/expectation.py [hunk 4]
```diff
@@ -1123,6 +1123,14 @@ def validate(
         if not configuration:
             configuration = deepcopy(self.configuration)
 
+        # issue warnings if necesary
+        self._warn_if_result_format_config_in_runtime_configuration(
```
Comment (anthonyburdi) [inline great_expectations/expectations/expectation.py:1160]: Love these descriptive method names!

File: tests/checkpoint/test_checkpoint_result_format.py [hunk 5]
```diff
@@ -308,30 +251,63 @@ def test_result_format_not_in_checkpoint_passed_into_run_checkpoint_one_expectat
...
         data_context=data_context_with_connection_to_animal_names_db,
-        checkpoint_config=reference_checkpoint_config_for_unexpected_column_names,
+        checkpoint_config=reference_sql_checkpoint_config_for_unexpected_column_names,
         expectations_list=[expectation_config_expect_column_values_to_be_in_set],
     )
-
     result_format: dict = {
-        "result_format": "COMPLETE",
+        "result_format": "SUMMARY",
+        "partial_unexpected_count": 1,
```
Comment (anthonyburdi) [inline tests/checkpoint/test_checkpoint_result_format.py:259]: :heart:

File: tests/expectations/metrics/column_map_metrics/test_unexpected_indices_and_query.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    unexpected_index_list = MetricConfiguration(
+        metric_name="column_values.in_set.unexpected_index_list",
+        metric_domain_kwargs={"column": "animals"},
+        metric_value_kwargs=metric_value_kwargs,
+    )
+    unexpected_index_list.metric_dependencies = {
+        "unexpected_condition": unexpected_columns_metric,
+        "table.columns": table_columns_metric,
+    }
+    with pytest.raises(MetricResolutionError):
```
Comment (anthonyburdi) [inline tests/expectations/metrics/column_map_metrics/test_unexpected_indices_and_query.py:228]: Ah took a minute but I understand now - this MetricResolutionError is raised if we are trying to retrieve the unexpected_index_list metric without specifying `unexpected_index_column_names`.

File: tests/expectations/metrics/test_map_metric.py [hunk 7]
```diff
@@ -349,6 +332,44 @@ def test_get_aggregate_count_aware_metric_dependencies(basic_spark_df_execution_
     )
 
 
+def test_get_aggregate_count_aware_metric_dependencies(basic_spark_df_execution_engine):
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_map_metric.py]: Is this a duplicate?

**Review by anthonyburdi — COMMENTED** [review]
Looks good, just one question about potential duplicate test.

File: tests/expectations/metrics/test_map_metric.py (hunk 7 above)
Comment (Shinnnyshinshin) [reply tests/expectations/metrics/test_map_metric.py]: Shoot you're right. Thanks for catching that @anthonyburdi

**Review by anthonyburdi — APPROVED** [review]
Thank you for this change and for the super helpful and descriptive PR description!

**Pull request merged**